### PR TITLE
upgrade job->payments data model

### PR DESCRIPTION
**Pull request by kategreenUSDS**
FFS-1227

### What's in here
* shifting to a more sophisticated data model for jobs and payments
* updating screens to reflect changes in the designs
* updating _other_ screens to align with data model changes (not in designs, but hewing as closely as i could)
* adding interpolation in the translation

### Screens and videos
Coming soon

### Reviewer requests
This is a big data model shift. 

### `TODO` before I remove the `DRAFT` tag

- [x] figure out how to test datepickers/whatever the hell is wrong in here `income/edit/[idx]/page.test.tsx:40`
- [x] update tests
- [x] lint
- [x] gut check assumptions on data structure; i did some fuzzy things that don't allow for assured types that need to be ironed out (see: `incomeSlice.ts` to see that `JobItem` has an optional `payments` type but it really should be optional and defined on create. stuff like that)
- [x] test out edit flows
- [x] lightly test end to end

### separate tickets
- [ ] update pdf
- [ ] add payment edit+delete functionality once there is designs for how to get there
- [ ] run through the flow after expenses tickets to check all designs
- [ ] snap monthly amount stub
- [ ] inconsistencies
  > there are a few inconsistencies in here right now that I'd like to handle eventually but in separate PRs. i'll cut a ticket with this for a spare moment. here are some things i've identified to iron out:
  > - standardize indentations+tabs+spaces
  > - align naming schema with translations and data model and test ids
  > - pre-push hooks to run npm run lint and npm run test
- [ ] functional flow tests

### CODE REVIEW

File: app/[locale]/ledger/income/IncomeFormPayment.tsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            </FormGroup>
+
+            <FormGroup>
+                <Controller
+                    name="date"
+                    control={control}
+                    rules={{ required: {value:true, message: t('add_income_required_field')} }}
+                    render={({ field }) => (
+                        <>
+                            <Label htmlFor="date">{t('add_income_payment_date')}<RequiredMarker /></Label>
```
Comment (joeyg) [inline app/[locale]/ledger/income/IncomeFormPayment.tsx:56]: The other labels have the `text-bold` css class.
Screenshot 2024-09-20 at 1 17 27 PM

File: app/components/IncomeListItem.tsx [hunk 2]
```diff
@@ -22,15 +25,22 @@ export default function IncomeListItem({ item, index }: ItemProps) {
...
+                <div>{item.business}</div>
+                <div>
+                    <ul>
+                        {payments}
+                    </ul>
+                </div>
             </Grid>
-            <Grid col={5} tablet={{col: 2}}>
+            <Grid col={2}>
                 <ModalToggleButton modalRef={ref} opener unstyled className="margin-bottom-2">{t('list_income_delete_button')}</ModalToggleButton>
```
Comment (joeyg) [inline app/components/IncomeListItem.tsx:44]: Not from this PR but there should probably be some space between these buttons. 
Screenshot 2024-09-20 at 1 20 14 PM

File: app/[locale]/ledger/income/IncomeFormJob.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+          error={errors.business?.message}
+          requiredMarker={false}
+          data-testid="business"
+          value={params.item?.business}
+        />
+      </FormGroup>
+
+      <FormGroup>
+        <legend className="usa-legend text-bold">{t('add_income_taxes')}</legend>
+        <Radio
```
Comment (joeyg) [inline app/[locale]/ledger/income/IncomeFormJob.tsx:80]: When editing a payment the value that was previously used does not start out checked. 
Screenshot 2024-09-20 at 1 23 54 PM

File: app/[locale]/ledger/income/edit/[idx]/page.tsx [hunk 4]
```diff
@@ -14,17 +14,17 @@ export default function EditIncome({ params }: { params: { idx: number } }) {
...
-                amount,
-            } as IncomeItem,
+                business,
+                taxesFiled
+            } as JobItem,
             idx: params.idx,
         }))
 
-        router.push('/ledger/income/list')
+        router.push(`/ledger/income/${params.idx}/payment/add`)
```
Comment (joeyg) [inline app/[locale]/ledger/income/edit/[idx]/page.tsx:27]: When editing a job the user should probably be redirected to the income list screen not the payment add screen.

**Review by joeyg — APPROVED** [review]
Looks great! I left a few comments inline and have two other questions:

- Is there a way to edit or remove payments?
- This wasn't part of the PR but the ledger review screen is a bit confusing and should probably be looked at:
Screenshot 2024-09-20 at 1 28 31 PM

-

File: app/[locale]/ledger/income/IncomeFormPayment.tsx (hunk 1 above)
Comment (kategreenUSDS) [reply app/[locale]/ledger/income/IncomeFormPayment.tsx:56]: ooh good catch

File: app/i18n/locales/en/translation.json [hunk 5]
```diff
@@ -70,6 +85,8 @@
   "expenses_snap_standard_title": "SNAP self employement expenses",
```
Comment (YvetteWhiteUSDS) [inline app/i18n/locales/en/translation.json:85]: Not part of your pr work, but there's a typo in employment.

File: app/i18n/locales/en/translation.json [hunk 6]
```diff
@@ -81,15 +98,15 @@
   "expenses_summary_title": "Expenses Summary",
   "expenses_summary_total": "Total expenses added: ${{amount}}",
   "expenses_summary_yes_delete_item": "Yes, remove this expense",
-  "how_this_works_header": "How this works",
-  "how_this_works_accordion_what_is_self_employment_title": "What is self-employed work?",
   "how_this_works_accordion_what_expenses_count_title": "What expenses count?",
+  "how_this_works_accordion_what_is_self_employment_title": "What is self-employed work?",
```
Comment (YvetteWhiteUSDS) [inline app/i18n/locales/en/translation.json:102]: This `self-employed` has a hyphen, but the others do not. Need to ask the design team if there should be a hyphen or not.

File: lib/store.ts [hunk 7]
```diff
@@ -28,15 +28,37 @@ export type AppStore = ReturnType<typeof makeStore>
...
     const percent = benefits.deductionAmount / 100
     return expenseTotal < incomeTotal * percent
+}
+
+/**
+ * Do we recommend they take a standard deduction based on SNAP?
+ * This is to evaluate before they add expenses
+ * 
+ * TODO: add amount that expenses should be less than
+ * TODO: add tests
```
Comment (YvetteWhiteUSDS) [inline lib/store.ts:55]: As with earlier comment about TODOs, considering adding an accompanying issue to make sure it happens.

**Review by YvetteWhiteUSDS — APPROVED** [review]

**Comment by kategreenUSDS** [comment]
@joeyg yes, payment editing hasn't yet been implemented. i am going to be writing another ticket for that. i don't want to get too far ahead of potential design changes coming in. there are some other data model change requests that may require some adjustments to the routing.

i'm going to tackle the spacing changes today and capture the expenses and payment changes in another ticket. 

at this point, i foresee a bunch of design changes coming, so i just want to get this merged and tighten things up as we handle other screen and data model shifts.

File: app/[locale]/ledger/income/IncomeFormJob.tsx (hunk 3 above)
Comment (kategreenUSDS) [reply app/[locale]/ledger/income/IncomeFormJob.tsx:80]: got it. thank you

File: app/[locale]/ledger/income/edit/[idx]/page.tsx (hunk 4 above)
Comment (kategreenUSDS) [reply app/[locale]/ledger/income/edit/[idx]/page.tsx:27]: at this point for sure.

File: app/[locale]/ledger/expense/snap/page.tsx [hunk 8]
```diff
@@ -1,114 +1,98 @@
...
     const { t } = useTranslation()
     const router = useRouter()
-    const dispatch = useAppDispatch()
     const benefits = useAppSelector(state => selectBenefits(state))
-    const expenseTotal = useAppSelector(state => selectExpenseTotal(state))
-    const reccommendStandardDeduction = useAppSelector(state => selectRecommendStandardDeduction(state))
+    const dispatch = useAppDispatch()
+
+    // todo: set this from state
+    const MONTHLY_AMOUNT = "XXX.XX"
```
[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/expense/snap/page.tsx:18]: yup. good catch.

File: app/[locale]/ledger/expense/snap/page.tsx [hunk 9]
```diff
@@ -1,114 +1,98 @@
...
+        let updatedBenefits = {...benefits}
+        updatedBenefits.standardDeduction = snapRadio
 
-            dispatch(setBenefits(newBenefits))
+        dispatch(setBenefits(updatedBenefits))
+
+        if (snapRadio == true) {
             router.push("/ledger/review")
+        } else if (snapRadio == false) {
+            router.push("/ledger/expense")
```
[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/expense/snap/page.tsx:39]: good find. suspect my text editor was trying to be "helpful" and i didn't catch it

[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/expense/snap/page.tsx:39]: could also become a ternary

File: app/[locale]/ledger/expense/snap/page.tsx [hunk 10]
```diff
@@ -1,114 +1,98 @@
...
         <div>
-            <VerifyNav title={t('snap_deduction_title')} />
+            <VerifyNav title={t('expenses_snap_standard_title')} />
             <div className="usa-section">
                 <GridContainer>
                     <Grid row gap>
                         <main className="usa-layout-docs">
                             <Form onSubmit={handleSubmit(onSubmit)}>
-                                <h3 className="margin-bottom-2" data-testid="snap_deduction_header">{t('snap_deduction_header', {amount: expenseTotal})}</h3>
+                                <h3 className="margin-bottom-2" data-testid="expense-snap-header">{t('expenses_snap_standard_header', {amount: MONTHLY_AMOUNT})}</h3>
```
[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/expense/snap/page.tsx:65]: habit. things in html get `-`s in my brain. i'll make it standard, though. 

for the record, these are my brian's defaults:
* js gets `camelCase`
* back end languages (and python no matter where) gets `under_scores_ftw`
* and css and other html-y things get `dashes-unless-its-bem` and then it's `this-thing__subclass` and `this-thing--variant`

[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/expense/snap/page.tsx:65]: and tbh i think i changed this one to match its translation key, which is html-y and has `-`s! 🙃

[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/expense/snap/page.tsx:65]: back in my fullstack days, this all helped my brain to context switch from backend to front end to styling, back when it was `php`+`jQuery`+`css` all separate

File: app/[locale]/ledger/expense/snap/recommend/page.tsx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        id: 'snap_accordion',
+        headingLevel: 'h4' as HeadingLevel,
+      }
+    ]
+
+    useEffect(() => {
+        if (!benefits.snap || !standardDeductionIsBetter) {
+            // User should not have been pushed to this screen
+            return router.push('/ledger/review')
+        }
```
[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/expense/snap/recommend/page.tsx:64]: This is a guard against a user being misdirected. It's legacy. One test type I'm thinking about is what we called "functional flow tests" on direct file: is the user being sent to the correct next step in the flow? This type of test would catch these errors (one hopes) and eliminate the need for this kind of guard. 

This isn't high on my list to implement, but i'm going to make a ticket that may end up in a gh issue here (cuz i'm going to make a few of them for contributors). cc: @joeyg

File: app/[locale]/ledger/income/IncomeFormJob.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  onSubmit: SubmitHandler<IncomeFormJobData>
+  item?: JobItem
+}
+
+export type IncomeFormJobData = {
+  description: string
+  business: string
+  taxesFiled: boolean
+  payments: []
+}
```
[the remark this answers is not in the document]
Comment (kategreenUSDS) [reply app/[locale]/ledger/income/IncomeFormJob.tsx:26]: fair point; the data model before the `Job->many Payments` notion included the word `Income`. I didn't do a systemic shift.

there are a few inconsistencies in here right now that I'd like to handle eventually but in separate PRs. i'll cut a ticket with this for a spare moment. here are some things i've identified to iron out:

1. standardize indentations+tabs+spaces
2. align naming schema with translations and data model and test ids
3. pre-push hooks to run `npm run lint` and `npm run test`

this is low priority when it's just us in here, but will become more important when we grow in dev involvement

File: lib/store.ts (hunk 7 above)
Comment (kategreenUSDS) [reply lib/store.ts:55]: yeah, agree. those notes were during a flurry of coding early this week.

**Pull request merged**